velox/exec/MergingVectorOutput.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/exec/MergingVectorOutput.h"
+#include "velox/exec/OperatorUtils.h"
+
+namespace facebook::velox::exec {
+
+MergingVectorOutput::MergingVectorOutput(
+    velox::memory::MemoryPool* pool,
+    int64_t preferredOutputBatchBytes,
+    int32_t preferredOutputBatchRows,
+    int32_t minOutputBatchRows)
+    : pool_(pool),
+      preferredOutputBatchBytes_(preferredOutputBatchBytes),
+      preferredOutputBatchRows_(preferredOutputBatchRows),
+      minOutputBatchRows_(minOutputBatchRows) {}
+
+RowVectorPtr MergingVectorOutput::getOutput() {
+  if (noMoreInput_) {
+    flush();
+  }
+  if (outputQueue_.size() > 0) {
+    const auto res = outputQueue_.front();
+    outputQueue_.pop();
+    return res;
+  }
+  return nullptr;
+}
+
+void MergingVectorOutput::addVector(RowVectorPtr vector) {
+  const auto numInput = vector->size();
+  if (numInput == 0) {
+    return;
+  }
+
+  // Avoid memory copying for pages that are big enough
+  if (numInput >= minOutputBatchRows_) {
+    outputQueue_.push(std::move(vector));
+    return;
+  }
+
+  // Currently, we only merge flat and dictionary vector, whether merging other
+  // encodings should depend on benchmark or adaptivity.
+  for (auto children : vector->children()) {
+    if (!VectorEncoding::isFlat(children->encoding()) &&
```

Comment:
Does that mean this optimization cannot be applies to data of complex types? (arrays, maps, structs)

Does this also mean that constant vectors disable this optimization?

I didn't see this restriction discussed in the PR description.

Reply:
@mbasmanova 
>Does that mean this optimization cannot be applies to data of complex types? (arrays, maps, structs)

The cost of merging each encoding is different. Memory copying of complex types may be more expensive than simple types. This requires additional benchmark comparison. Currently the benchmark only includes flat and dictionary encoding tests.

> Does this also mean that constant vectors disable this optimization?

After the constant vectors are merged, they are no longer constant vectors, which may cause the downstream operator to process the constant vector slowly.

> I didn't see this restriction discussed in the PR description.

Sorry, I will add this part to the design description later.

Reply:
> The cost of merging each encoding is different. Memory copying of complex types may be more expensive than simple types. This requires additional benchmark comparison. Currently the benchmark only includes flat and dictionary encoding tests.

Hmm... notice that dictionary encoding may be applied to complex type vector, in which case you will be copying complex types.

Reply:
The most critical point is how to measure the copy cost and merger benefits of complex types. A similar simple solution is to add benchmarks for complex types and then give recommended minMergingVectorOutputBatchRows.

Reply:
Indeed. Maybe you can do more benchmarking to see if a fixed config value can work work well in most cases. If not, we may need to think about how to enable/disable this functionality adaptively.

Also, we may want to identify operators that require min batch size for inputs and have some logic in LocalPlanner to enable merging output only for operators that produce data to these.